src/concrete/ml/quantization/quantized_ops.py

```diff
@@ -217,20 +225,163 @@ def q_impl(
         # Here we follow Eq.7 in https://arxiv.org/abs/1712.05877 to split the core computation
         # from the zero points and scales.
 
-        p = weights_q_values.shape[0]
+        p = input2_q_values.shape[-2]
+
+        # Remove the manual matrix multiplication when we can handle input precision with rounding
+        # FIXME: https://github.com/zama-ai/concrete-internal/issues/512
+        def enc_mul(x, y):
+            r"""Encrypted multiplication of two input arrays.
+
+            This function computes the encrypted multiplication of two numpy arrays.
+            It uses the following equality:
+
+            \[
+            (x + y)^2 - (x - y)^2 = 4xy
+            \]
+
+            This equation simplifies to the standard multiplication operation.
+
+            In TFHE, this allows to do encrypted multiplication with 2 PBS.
+
+            Args:
+                x (numpy.ndarray): The first input numpy array.
+                y (numpy.ndarray): The second input numpy array.
+
+            Returns:
+                numpy.ndarray: The result of the encrypted multiplication.
+            """
+            with tag("pbs_multiplication"):
+                # Compute sum and difference of x and y
+                add = x + y
+                sub = x - y
+
+                # Apply Concrete rounding to the addition and substraction
+                with tag(self.op_instance_name + ".pbs_matmul_rounding_add"):
+                    add = self.cnp_round(add, calibrate_rounding, rounding_operation_id="add")
+                with tag(self.op_instance_name + ".pbs_matmul_rounding_sub"):
+                    sub = self.cnp_round(sub, calibrate_rounding, rounding_operation_id="sub")
+
+                # Square the rounded sums and differences, and divide by 4
+                add_pow = (add.astype(numpy.float64)) ** 2
+                sub_pow = (sub.astype(numpy.float64)) ** 2
+                add_pow_divide = (add_pow / 4.0).astype(numpy.int64)
```

Comment:
would be better to do: (add_pow - sub_pow) / 4, to have only one div ?

Reply:
We can't do that. Here we cast to floating points so both values are inputs to a PBS. A PBS can't have 2 encrypted inputs. So here we are basically building two different PBS.